### PR TITLE
Bioconda delay analysis

### PR DESCRIPTION
**Pull request by luispedro**
As discussed on authorea, this looks at the delay between the upstream release and the bioconda package.

1. Look for the first commit that introduced the current version of the package. This is the bioconda date.
2. Attempt to infer the upstream date: if it's a github/pypi release, use the metadata there. Otherwise, download the file and look at mtimes.

This does not work for all packages, but it works for >90% of them. Packages that refer to a particular git commit are ignored (I can also change the code to use the date of that commit).

I then summarized the results (while ignoring upstream packages from before 2016). These are the results I get:

    Mean (+/- std.dev.) number of days between upstream release and bioconda package: 102.122157245 +/- 101.787918729
    Median number of days between upstream release and bioconda package: 98.0
    Based on 1539 packages.

    On those packages where it could be heuristically determined, 228 of 358 are current with their upstream release.

(If this gets incorporated into the paper, then I would like to be an author [Luis Pedro Coelho, Structural and Computational Biology Unit, European Molecular Biology Laboratory, Heidelberg, Germany]. I have also added a recent package to bioconda, so I guess that would also qualify me, except I might have missed that deadline).

As discussed, I also got a (very partial) list of which packages are outdated wrt their upstream releases.

*

BTW, I wasn't sure at all where to put these scripts in this repo. This seemed the most logical place, but let me know if you prefer me to move them.

### CODE REVIEW

**Comment by johanneskoester** [comment]
This is very promising, thanks! In order to get unbiased results, it would be important to only consider packages where the current version is not the first one. The reason is that we only want to measure the delay for a new version to be picked if the package is already in the repo. Otherwise, we get arbitrary biases by the delay between initial inclusion and the last release of the software.

File: scripts/update-delay/recipe_times.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+            continue
+        try:
+            return int(ea) > int(eb)
+        except:
+            return ea > eb
+    return False
+
+def extract_meta(meta):
+    '''Tries to parse the contents of a meta.yaml file to build a RecipeInfo
+    object'''
```
Comment (johanneskoester) [inline scripts/update-delay/recipe_times.py:31]: You can import conda for parsing the meta.yaml. No need to re-implement this.

File: scripts/update-delay/recipe_times.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    time.sleep(.1) # Do not bombard the server
+    if 'GITHUB_USERNAME' not in environ or \
+        'GITHUB_TOKEN' not in environ:
+        from sys import exit
+        print("Please set the environmental variables GITHUB_USERNAME and GITHUB_TOKEN to query Github's API")
+        exit(1)
+
+    what = ('release' if not fetch_tags else 'tags')
+    r = requests.get('https://api.github.com/repos/{}/{}/{}'.format(owner, repo, what),
+                auth=(environ['GITHUB_USERNAME'], environ['GITHUB_TOKEN']))
```
Comment (johanneskoester) [inline scripts/update-delay/recipe_times.py:170]: You can use pygithub to access the github API. Please see my collect-pr-data script: https://github.com/bioconda/bioconda-paper/blob/master/scripts/collect-pr-data.py

**Review by johanneskoester — CHANGES_REQUESTED** [review]

**Comment by luispedro** [comment]
You are right that we need to be careful with older packages. I had set an arbitrary cut-off for packages earlier than 2016, but you are right that only considering packages with >1 bioconda version is a better system. I will reimplement it like that.

(I will also fix the other issues).

**Comment by luispedro** [comment]
Updated results (now only considering packages with >1 version on bioconda):

    Mean (+/- std.dev.) number of days between upstream release and bioconda package: 83.1293103448 +/- 99.1218653899
    Median number of days between upstream release and bioconda package: 37.0
    Based on 696 packages.

    On those packages where it could be heuristically determined, 160 of 217 are current with their upstream release.

File: scripts/update-delay/recipe_times.py (hunk 1 above)
Comment (luispedro) [reply scripts/update-delay/recipe_times.py:31]: I've now spent more time trying to figure out how to update conda-build on the right environment than I had spent reimplementing this.

**Comment by luispedro** [comment]
Following-up on this (and this recent piece of info: https://github.com/bioconda/bioconda-recipes/issues/6323#issuecomment-347994955):

Is this OK now? The big issue of which packages to count has been solved.

**Review by johanneskoester — APPROVED** [review]

**Comment by johanneskoester** [comment]
Yes, I think it is fine. I will merge it and we will keep it in mind for the first revision.

**Pull request merged**